pkg/kube/kube_handler_test.go

```diff
@@ -1088,137 +1093,523 @@ func setClientsetData(cs fakeCombinedClientset, namespaceNames []existingNs, err
+			httpValidator:    HelmNonOCIValidator{Req: validRequest},
 		},
 	}
 
-	for _, tc := range getValidationCliAndReqTests {
+	cmpOpts := []cmp.Option{cmpopts.IgnoreUnexported(http.Request{}, strings.Reader{})}
+	cmpOpts = append(cmpOpts, cmpopts.IgnoreFields(http.Request{}, "GetBody"))
+
+	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {
-			cs := fakeCombinedClientset{
-				fakeapprepoclientset.NewSimpleClientset(),
-				fakecoreclientset.NewSimpleClientset(),
-				&fakeRest.RESTClient{},
-			}
-			handler := userHandler{
-				kubeappsNamespace: kubeappsNamespace,
-				svcClientset:      cs,
-				clientset:         cs,
-			}
-			appRepo, cli, err := handler.getValidationCli(ioutil.NopCloser(strings.NewReader(tc.requestData)), tc.requestNamespace, kubeappsNamespace)
-			if (err != nil || tc.expectedError != nil) && !errors.Is(err, tc.expectedError) {
-				t.Fatalf("got: %+v, want: %+v", err, tc.expectedError)
+			fakeClient := &fakeHTTPCli{
+				response: tc.fakeRepoResponse,
+				err:      tc.fakeHttpError,
 			}
-			if tc.expectedError != nil {
-				return
+
+			response, err := tc.httpValidator.Validate(fakeClient)
+			if err != nil {
+				t.Errorf("Unexpected error %v", err)
 			}
-			httpValidator, err := getValidator(appRepo, cli)
-			if (err != nil || tc.expectedReqResolutionError != nil) && !errors.Is(err, tc.expectedReqResolutionError) {
-				t.Fatalf("got: %+v, want: %+v", err, tc.expectedReqResolutionError)
+
+			if got, want := fakeClient.request, tc.httpValidator.Req; !cmp.Equal(want, got, cmpOpts...) {
+				t.Errorf("mismatch (-want +got):\n%s", cmp.Diff(want, got, cmpOpts...))
 			}
-			if tc.expectedReqResolutionError != nil {
-				return
+
+			if got, want := response, tc.expectedResponse; !cmp.Equal(want, got) {
+				t.Errorf("mismatch (-want +got):\n%s", cmp.Diff(want, got))
 			}
-			if appRepo.Spec.Type != "oci" {
+		})
+	}
+}
 
-				req := httpValidator.getReq()
-				reqURLs := []string{}
-				reqURLs = append(reqURLs, req.URL.String())
+type fakeOCIRepo struct {
+	tags     repoTagsList
+	manifest repoManifest
+}
 
-				if !cmp.Equal(tc.expectedURLs, reqURLs) {
-					t.Errorf("Unexpected URLS: %v", cmp.Diff(tc.expectedURLs, reqURLs))
-				}
-				if tc.expectedHeaders != nil && !cmp.Equal(tc.expectedHeaders, cli.(*httpclient.ClientWithDefaults).DefaultHeaders) {
-					t.Errorf("mismatch (-want +got):\n%s", cmp.Diff(tc.expectedHeaders, cli.(*httpclient.ClientWithDefaults).DefaultHeaders))
-				}
-			} else {
-				httpValidatorAppRepo := httpValidator.getAppRepo()
-				if httpValidatorAppRepo != appRepo {
-					t.Errorf("Unexpected Repos: %v", cmp.Diff(appRepo, httpValidatorAppRepo))
-				}
+// makeTestOCIServer returns a small test double for an OCI server that handles requests
+// for tags/list and manifest only.
+func makeTestOCIServer(t *testing.T, registryName string, repos map[string]fakeOCIRepo, requiredAuthHeader string) *httptest.Server {
```

Comment:
Reality is, it's not great to only unit-test the internal functions, we want to be able to unit test the actual exposed handler as well. This wasn't possible for the OCI success case (which I guess is why you didn't get to add them :) ) without either a massive refactoring to enable injecting fake clients, *or* by creating a test double for the upstream OCI server, which is the path I've gone for here. For a given registry and set of repos, it responds to the `tag/list` and `manifest` endpoints only: